File: Data Structure/C/binary_search_tree.c
//function:create_new_node,search,insert,delete,find_maximum,find_minimum,find_successor;

struct TreeNode {
     int val;
     struct TreeNode *left;
     struct TreeNode *right;
};

//search
struct TreeNode* searchBST(struct TreeNode* root, int val){
    while(root!=NULL && val!=root->val){
        if(val<root->val)
            root=root->left;
        else
            root=root->right;
    }
    return root;
}

//create a new node
struct TreeNode* new_node(int key){
    struct TreeNode* node=(struct TreeNode*)malloc(sizeof(struct TreeNode));
    node->val=key;
    node->left=node->right=NULL;
    return node;
}

//insert
struct TreeNode* insertIntoBST(struct TreeNode* root, int val){
    struct TreeNode* node=new_node(val);
    struct TreeNode* tmp=root;
    struct TreeNode* par;
    if(root==NULL)
        return node;
    while(tmp!=NULL){
        if(val<tmp->val){
            par=tmp;
            tmp=tmp->left;
        }
        else{
            par=tmp;
            tmp=tmp->right;
        }
    }
    if(val<par->val)
        par->left=node;
    else
        par->right=node;
    return root;
}

//find minimum in a subtree
struct TreeNode* minimum(struct TreeNode* node){
        while(node->left)
                node = node->left;
        return node;
}

//find maximum in a subtree
struct TreeNode* maximum(struct TreeNode* node){
        while(node->right)
                node = node->right;
        return node;
}

//find successor of a node
struct TreeNode* succeed(struct TreeNode* node){
        if(!(node->right))
                return maximum(node->left);
          else
                return minimum(node->right);
}

//input a key
//delete the node with that key in the bst
//in a iterative way
struct TreeNode* deleteNode(struct TreeNode* root, int key){
        if(!root)
                return NULL;
        struct TreeNode *del,*par,*suc,*tmp;
        par = root;
        del = root;
        while(del && del->val != key){
                par=del;
                if(key < del->val)
                        del = del->left;
                else
                        del = del->right;
        }
        if(!del) return root;                 //no node to be deleted
        if( !(del->left) && !(del->right)){           //node have no child
                if(del->val == root->val)
                        return NULL;              //delete root
                if(del == par->left).            //delete leaf
                          par->left = NULL;
                  else
                          par->right = NULL;
                  return root;
          }
     if(del->left && !(del->right)){.       //node only have l. subtree
                if(del == root)
                        return root->left;
                if(del == par->left)
                        par->left = del->left;
                else
                        par->right = del->left;         //skip
        }
        else if(del->right && !(del->left)){.       //node only have r. subtre  e
                if(del == root)
                        return root->right;
                if(del == par->left)
                        par->left = del->right;
                else
                        par->right = del->right;       //skip
        }
         else{                       //node have two subtrees
                suc = succeed(del);
                if(del->right == suc){               //del->right->left == NUL  L
                        if(del == root){
                                suc->left = del->left;    //succeed root'  s l. subtree
                                return suc;
                        }
                        if(del == par->left){
                                par->left = suc; //skip
                        }
                        else
                                par->right = suc; //skip
                        suc->left = del->left; //succeed node's l. subtree
                }
             else{   //normal case
                        del->val = suc->val;
                        tmp = del->right;
                        while(tmp->left && tmp->left != suc)
                                tmp = tmp->left;     //find successor's parent
                        if(suc->right)
                                tmp->left =suc->right;
                        else                //successor has no child;
                        tmp->left=NULL;
                }
        }
        return root;
}

//input a key
//delete the node with that key in the bst
//in a recursion way
struct TreeNode* deleteNode_recur(struct TreeNode* root, int key){
       if (!root)
                return NULL;
        if(key<root->val)
                root->left=deleteNode(root->left,key);
        else if(key>root->val)
                root->right=deleteNode(root->right,key);
        else{
            if(!(root->left)){
                struct TreeNode* base=root->right;
                root = NULL;
                return base;
            }
            else if(!(root->right)){
                struct TreeNode* base=root->left;
                root = NULL;
                return base;
            }
            else{
                struct TreeNode* base=minimum(root->right);
                root->val=base->val;
                root->right=deleteNode(root->right,base->val);
            }
        }
    return root;
}

File: Data Structure/GO/binarySearchTree.go
//
package main

import "fmt"

//This is a binary search tree
//All the values in the left subtree is smaller than the root
//All the values in the right subtree are greater than the root

type treeNode struct {
	data   int
	left   *treeNode
	right  *treeNode
	parent *treeNode
}

type binarySearchTree struct {
	root *treeNode
}

func newTreeNode(t int) *treeNode {
	return &treeNode{data: t}
}

func newBinarySearchTree() *binarySearchTree {
	return &binarySearchTree{}
}

//Insert a leaf into a tree
func (tree *binarySearchTree) push(t int) {
	n := newTreeNode(t)
	var tmp *treeNode = nil
	walker := tree.root
	for walker != nil {
		tmp = walker
		if t < walker.data {
			walker = walker.left
		} else {
			walker = walker.right
		}
	}
	n.parent = tmp
	if tmp == nil {
		tree.root = n
		return
	}
	if t < tmp.data {
		tmp.left = n
	} else {
		tmp.right = n
	}
}

//the iteration stop when the walker walk out of the tree
//or the walker walk to the node we want
func (tree *binarySearchTree) search(key int) *treeNode {
	walker := tree.root
	for walker != nil && walker.data != key {
		if key < walker.data {
			walker = walker.left
		} else {
			walker = walker.right
		}
	}
	return walker
}

func (node *treeNode) predecessor() *treeNode {
	pre := node.left
	for pre != nil && pre.right != nil {
		pre = pre.right
	}
	return pre
}

func (node *treeNode) successor() *treeNode {
	pre := node.right
	for pre != nil && pre.left != nil {
		pre = pre.left
	}
	return pre
}

//the case following are disjoint
//case 1:no node to be deleted
//case 2:the node to be deleted is a leaf
//case 3:the node to be deleted has both predecessor and successor
//case 4:the node to be deleted has either successor or predecessor
//successor and predecessor has no subtree,it is a leaf
func (tree *binarySearchTree) delete(key int) {
	n := tree.search(key)
	if n == nil { //case 1
		return
	}
	if n.left == nil && n.right == nil { //case 2
		n = nil
		return
	}
	var suc *treeNode
	if n.right == nil { //case 4
		suc = n.predecessor()
	} else { //case 3 or 4
		suc = n.successor()
	}
	if n == tree.root { //those case will change the root aritribute
		if n.left == nil && n.right == nil {
			tree.root = nil
			return
		}
		if suc == n.left {
			suc.right = n.right
			tree.root = suc
		}
		if suc == n.right {
			suc.left = n.left
			tree.root = suc
		}
	}
	n.data = suc.data
	if suc == suc.parent.left {
		suc.parent.left = nil
	} else {
		suc.parent.right = nil
	}
}

func (node *treeNode) inorder() {
	if node != nil {
		node.left.inorder()
		fmt.Printf("%d ", node.data)
		node.right.inorder()
	}
}

func (tree *binarySearchTree) inorderTraversal() {
	tree.root.inorder()
	fmt.Println()
}

func main() {
	tree := newBinarySearchTree()
	tree.push(5)
	tree.push(7)
	tree.push(4)
	tree.push(6)
	tree.push(3)
	tree.push(1)
	tree.inorderTraversal()
	tree.delete(5)
	tree.inorderTraversal()
	tree.delete(5)
	tree.inorderTraversal()
	tree.delete(6)
	tree.delete(0)
	tree.inorderTraversal()
}
